File: neetcode150-py/src/neetcode150_py/arrays_and_hashing/q128_longest_consecutive_sequence.py
from typing import List, Dict


class Solution:
    """
    Constraint: T(n) = O(n)
    """

    def longestConsecutive(self, nums: List[int]) -> int:
        """
        T(n) = O(n)
        S(n) = O(n)
        """
        longest = 0
        nums_set = set(nums)

        for n in nums:
            if n - 1 in nums_set:  # It's not the leftmost number in the sequence
                continue

            cur_n = n
            while cur_n in nums_set:
                cur_n += 1
            longest = max(cur_n - n, longest)

        return longest

    def longestConsecutive1(self, nums: List[int]) -> int:
        """
        T(n) = O(n^2)
        S(n) = O(2n) = O(n)
        """
        if len(nums) == 0:
            return 0

        table: Dict[int, List] = {}
        nums_set = set(nums)
        for n in nums:
            cur_n = n
            while cur_n - 1 in nums_set:
                cur_n -= 1
            if cur_n not in table or (
                cur_n in table and len(table[cur_n]) < n + 1 - cur_n
            ):
                table[cur_n] = list(range(cur_n, n + 1))

        return max(len(x) for x in table.values())

File: neetcode150-rs/src/arrays_and_hashing/q128_longest_consecutive_sequence.rs
use std::collections::HashSet;

pub struct Solution;

impl Solution {
    pub fn longest_consecutive(nums: Vec<i32>) -> i32 {
        let set: HashSet<i32> = HashSet::from_iter(nums);
        let mut max_seq_len = 0;

        for &num in &set {
            if set.contains(&(num - 1)) {
                // this num is not a seq head
                continue;
            }
            // this num is a seq head, keep searching up
            let mut cur_seq_len = 1;
            while set.contains(&(num + cur_seq_len)) {
                cur_seq_len += 1;
            }

            max_seq_len = max_seq_len.max(cur_seq_len);
        }
        max_seq_len
    }
}

#[cfg(test)]
mod tests {
    use crate::arrays_and_hashing::q128_longest_consecutive_sequence::Solution;

    /// Example 1:
    /// Input: nums = [100,4,200,1,3,2]
    /// Output: 4
    /// Explanation: The longest consecutive elements sequence is [1, 2, 3, 4]. Therefore its length is 4.
    #[test]
    fn test_longest_consecutive_example_1() {
        let nums = vec![100, 4, 200, 1, 3, 2];
        assert_eq!(Solution::longest_consecutive(nums), 4);
    }

    /// Example 2:
    /// Input: nums = [0,3,7,2,5,8,4,6,0,1]
    /// Output: 9
    #[test]
    fn test_longest_consecutive_example_2() {
        let nums = vec![0, 3, 7, 2, 5, 8, 4, 6, 0, 1];
        assert_eq!(Solution::longest_consecutive(nums), 9);
    }

    /// Example 3:
    /// Input: nums = [1,0,1,2]
    /// Output: 3
    #[test]
    fn test_longest_consecutive_example_3() {
        let nums = vec![1, 0, 1, 2];
        assert_eq!(Solution::longest_consecutive(nums), 3);
    }
}
